### PR TITLE
[REVIEW] Fix get_dummies to be compatible with dask

### PR DESCRIPTION
**Pull request by galipremsagar**
- [x] Dask currently passes default dtype `uint8`: https://github.com/dask/dask/blob/master/dask/dataframe/reshape.py#L25
which cause below exception to occur:

```python
KeyError: <class 'numpy.uint8'>
Exception ignored in: 'cudf.bindings.cudf_cpp.column_view_from_column'
KeyError: (<class 'numpy.uint8'>,)
KeyError: <class 'numpy.uint8'>
Exception ignored in: 'cudf.bindings.cudf_cpp.column_view_from_column'
KeyError: (<class 'numpy.uint8'>,)
KeyError: <class 'numpy.uint8'>
Exception ignored in: 'cudf.bindings.cudf_cpp.column_view_from_column'
KeyError: (<class 'numpy.uint8'>,)
KeyError: <class 'numpy.uint8'>
Exception ignored in: 'cudf.bindings.cudf_cpp.column_view_from_column'
KeyError: (<class 'numpy.uint8'>,)
KeyError: <class 'numpy.uint8'>
Exception ignored in: 'cudf.bindings.cudf_cpp.column_view_from_column'
KeyError: (<class 'numpy.uint8'>,)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/dask/dask/base.py", line 175, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/dask/dask/base.py", line 447, in compute
    return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
  File "/dask/dask/base.py", line 447, in <listcomp>
    return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
  File "/conda/envs/cudf/lib/python3.7/site-packages/dask_cudf/core.py", line 48, in finalize
    return cudf.concat(results)
  File "/conda/envs/cudf/lib/python3.7/site-packages/cudf/multi.py", line 54, in concat
    return DataFrame._concat(objs, axis=axis, ignore_index=ignore_index)
  File "/conda/envs/cudf/lib/python3.7/site-packages/cudf/dataframe/dataframe.py", line 1447, in _concat
    for c in objs[0].columns
  File "/conda/envs/cudf/lib/python3.7/site-packages/cudf/dataframe/dataframe.py", line 1447, in <listcomp>
    for c in objs[0].columns
  File "/conda/envs/cudf/lib/python3.7/site-packages/cudf/dataframe/series.py", line 1034, in _concat
    col = Column._concat([o._column for o in objs])
  File "/conda/envs/cudf/lib/python3.7/site-packages/cudf/dataframe/column.py", line 112, in _concat
    col = _column_concat(objs, col)
  File "cudf/bindings/concat.pyx", line 29, in cudf.bindings.concat._column_concat
  File "cudf/bindings/cudf_cpp.pyx", line 487, in cudf.bindings.cudf_cpp.check_gdf_error
cudf.bindings.GDFError.GDFError: b'GDF_DATASET_EMPTY'
```
Fixed the above case by explicitly overriding to int8 for now, but this will have to go away once we start supporting `uint8`.

- [x] Added picking of categories from the categories column instead of calculating uniques again, which will bring `dask.dataframe._meta` pre-calculation of columns too to be happening correctly.

- [x] Added unit-tests for dask-cudf which should start passing once this PR of dask gets mergerd: https://github.com/dask/dask/pull/5090

cc: @beckernick @randerzander @kkraus14

### CODE REVIEW

File: python/dask_cudf/dask_cudf/tests/test_dataframe.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import pandas as pd
+
+import cudf
+import dask.dataframe as dd
+import dask_cudf
+
+import pytest
+
+
+def test_get_dummies(data):
```
Comment (beckernick) [inline python/dask_cudf/dask_cudf/tests/test_dataframe.py]: Tests for get_dummies currently live in `tests/test_onehot.py`. For consistency, we should include these there for now.

More broadly, @shwina , given your recent thinking about the directory structure, do you want to draw an organizational line in the in the sand between cudf and dask tied tests in the short/medium term?

File: python/dask_cudf/dask_cudf/tests/test_dataframe.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    df["B"] = df["A"].astype('category')
+    df["A"] = df["A"].astype('category')
+    ddf = dd.from_pandas(df, npartitions=10)
+    dd.assert_eq(dd.get_dummies(ddf).compute(), pd.get_dummies(df))
+    gdf = cudf.from_pandas(df)
+    gddf = dask_cudf.from_cudf(gdf, npartitions=10)
+    dd.assert_eq(dd.get_dummies(ddf).compute(), dd.get_dummies(gddf),
+                 check_dtype=False)
+
+    df = pd.DataFrame({"A": ["a", "b", "c", "a", "z"],
```
Comment (beckernick) [inline python/dask_cudf/dask_cudf/tests/test_dataframe.py]: You should include one example with a decent amount of data and significant number of partitions. You've passed 10 in all of these, but I believe this won't actually create 10 partitions given the data size. Sometimes, wonky behavior can surface by going down a codepath that only exposes itself with enough partitions.

File: python/dask_cudf/dask_cudf/tests/test_dataframe.py (hunk 1 above)
Comment (galipremsagar) [reply python/dask_cudf/dask_cudf/tests/test_dataframe.py]: sure, would move them to `test_onehot.py`. Will wait for @shwina inputs if any other organizational structure is recommended?

File: python/dask_cudf/dask_cudf/tests/test_dataframe.py (hunk 2 above)
Comment (galipremsagar) [reply python/dask_cudf/dask_cudf/tests/test_dataframe.py]: Added one example of a large dataset across 100 partitions.

File: python/dask_cudf/dask_cudf/tests/test_dataframe.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    df = pd.DataFrame({"C": [], "A": []})
+    ddf = dd.from_pandas(df, npartitions=10)
+    dd.assert_eq(dd.get_dummies(ddf).compute(), pd.get_dummies(df))
+    gdf = cudf.from_pandas(df)
+    gddf = dask_cudf.from_cudf(gdf, npartitions=10)
+    dd.assert_eq(dd.get_dummies(ddf).compute(), dd.get_dummies(gddf),
+                 check_dtype=False)
+
+    gdf = cudf.datasets.randomdata(nrows=200000, dtypes={"C": int,
```
Comment (kkraus14) [inline python/dask_cudf/dask_cudf/tests/test_dataframe.py]: Given this is a unit test where we have concerns about runtime, can we make sure we don't introduce long-running tests? How long does this section of the test currently take to run?

File: python/dask_cudf/dask_cudf/tests/test_dataframe.py (hunk 1 above)
Comment (kkraus14) [reply python/dask_cudf/dask_cudf/tests/test_dataframe.py]: @galipremsagar can you break this test out into multiple tests instead? That way if something fails we have a much more directed area to investigate.

Comment (shwina) [reply python/dask_cudf/dask_cudf/tests/test_dataframe.py]: @galipremsagar 

`dask_cudf/tests/test_onehot.py` sounds good to me.

Organizationally, as @beckernick suggests, I do think it makes sense to draw a line between cudf tests and dask tests. Tests for `cudf` live under the `cudf` package and tests for `dask_cudf` under the `dask_cudf` package. We should never have to `import dask_cudf` from anywhere within the `cudf` test suite.

File: python/dask_cudf/dask_cudf/tests/test_dataframe.py (hunk 3 above)
Comment (shwina) [reply python/dask_cudf/dask_cudf/tests/test_dataframe.py]: +1. If the concern is that a small data size may not result in enough partitions, then we need to figure out:

* the smallest number of partitions required to fully test correctness
* the smallest number of elements per partition required to fully test correctness 

- and figure out how to force the creation of such a dataset. Unit tests must be as small and as fast as possible.

Comment (galipremsagar) [reply python/dask_cudf/dask_cudf/tests/test_dataframe.py]: `test_get_dummies_large` (https://github.com/rapidsai/cudf/pull/2235/files#diff-bf8370b82b3eaa5737f32322702ad78eR61) takes 10 seconds to run.

Comment (kkraus14) [reply python/dask_cudf/dask_cudf/tests/test_dataframe.py]: So given there's 100 partitions, what if we reduced the nrows down to say 2000 and got ~20 rows per partition? That way in theory the runtime drastically reduces and we still get our 100 partition test.

Comment (galipremsagar) [reply python/dask_cudf/dask_cudf/tests/test_dataframe.py]: Actually reducing number of partitions helps:

200000 records - 100 partitions - 45 seconds
**200000 records - 25 partitions - 13 seconds** <- current configuration
2000 records - 100 partitions - 45 seconds
2000 records - 25 partitions - 11 seconds
200 records - 25 partitions - 11 seconds

File: python/dask_cudf/dask_cudf/tests/test_dataframe.py (hunk 1 above)
Comment (galipremsagar) [reply python/dask_cudf/dask_cudf/tests/test_dataframe.py]: @kkraus14 , @shwina refactored accordingly.

File: python/dask_cudf/dask_cudf/tests/test_dataframe.py (hunk 3 above)
Comment (galipremsagar) [reply python/dask_cudf/dask_cudf/tests/test_dataframe.py]: One more set of results:
200000 records - 10 partitions - 8 seconds
20000 records - 10 partitions - 5 seconds <- I think this looks good ?? @beckernick @kkraus14

Comment (kkraus14) [reply python/dask_cudf/dask_cudf/tests/test_dataframe.py]: 5 seconds still sounds extremely slow, no? What's taking so long in the test?

Comment (galipremsagar) [reply python/dask_cudf/dask_cudf/tests/test_dataframe.py]:
```python
In [10]: %time x = dd.get_dummies(gddf).compute()
CPU times: user 4.38 s, sys: 184 ms, total: 4.56 s
Wall time: 4.56 s

In [11]: %time x = dd.get_dummies(ddf).compute()
CPU times: user 60 ms, sys: 12 ms, total: 72 ms
Wall time: 68 ms
```

gddf being dask-cudf dataframe, ddf being dask dataframe.

Comment (kkraus14) [reply python/dask_cudf/dask_cudf/tests/test_dataframe.py]: Yea, something feels extremely off here, any chance you can generate a snakeviz profile before we merge this?

Comment (galipremsagar) [reply python/dask_cudf/dask_cudf/tests/test_dataframe.py]: Sending this offline.

Comment (kkraus14) [reply python/dask_cudf/dask_cudf/tests/test_dataframe.py]: Connected offline, there's performance concerns but they're somewhat unavoidable until we move more of this function down to the libcudf level.

@galipremsagar will open an issue to further discuss

Comment (galipremsagar) [reply python/dask_cudf/dask_cudf/tests/test_dataframe.py]: created an issue related to this: https://github.com/rapidsai/cudf/issues/2246

**Comment by galipremsagar** [comment]
rerun tests

**Pull request merged**